### DISCUSSION TITLE
Dynamic Background Color in Next.js with Tailwind CSS: Issue and Solution

### DISCUSSION THREAD

**Discussion by ajaymahadeven** [tailwindlabs/tailwindcss#13714] [Help]
Certainly! Here is the updated professional message including the explanation for the `getStatusCode` function and the relevant data structures:

---

### Issue Description

I encountered an issue while working on a Next.js project using TypeScript. The goal was to dynamically set the background color of a component based on a status ID. Here is the initial implementation:

```typescript
const statusColor = getStatusColor(status_id); // e.g., #ffff00, #ff9900, #ff0000, #33ff33
<div className="flex flex-col">
    <div
        className={`flex h-14 items-center justify-center space-x-4 rounded-md border bg-[${statusColor}] p-4 text-white`}
    >
        <div className="flex-1 space-y-1"></div>
    </div>
</div>
```

However, this approach did not work as expected when running `npm run dev` in a Next.js environment. The dynamic class for the background color was not applied correctly.

### Observations

1. **Initial Attempt**: The class containing the dynamic color value (`bg-[${statusColor}]`) did not apply the background color as intended during development.
2. **Temporary Workaround**: Changing the dynamic property to a fixed value and then reverting it back to the dynamic value seemed to work temporarily. However, this workaround only lasted until the server was restarted.
3. **Current Solution**: To address this issue, I applied the background color using the `style` attribute instead of relying on dynamic Tailwind CSS classes.

### Revised Code

```typescript
const statusColor = getStatusColor(status_id);
<div className="flex flex-col">
    <div
        className="flex h-14 items-center justify-center space-x-4 rounded-md border p-4 text-white"
        style={{ backgroundColor: statusColor }}
    >
        <div className="flex-1 space-y-1"></div>
    </div>
</div>
```

By forcefully setting the background color via the `style` prop, the color is assigned correctly during compilation.

### Explanation of `getStatusCode` Function

The `getStatusColor` function retrieves the color associated with a given status ID from a predefined list of status descriptions. Here is the relevant code:

```typescript
import { statusInfo } from '@/data/status-descriptions/status-descriptions';

export function getStatusColor(status_id: string) {
    const defaultColor = 'black';
    const status = statusInfo.find((status) => status.key === status_id);
    console.log('Status:', status);
    const statusColor: string = status ? status.color : defaultColor;

    return statusColor;
}
```

### Status Info Code

The `statusInfo` array is generated from a detailed description of various statuses:
```typescript

export const STATUS_DESCRIPTIONS: Record<string, Status> = {
    green: {
        color: '#33ff33',
        description: {
            lang: 'en',
            text: 'No significant activity',
        },
        meaning: {
            lang: 'en',
            text: 'Aurora is unlikely to be visible by eye or camera from anywhere in the UK.',
        },
    },
  {
  ...
  }
}

export const statusInfo = Object.entries(STATUS_DESCRIPTIONS).map(
    ([key, { color, description, meaning }]) => ({
        key,
        color,
        description: description.text,
        meaning: meaning.text,
    }),
);
```

### Conclusion

This solution ensures that the background colour is applied correctly irrespective of the server's state. Although using the `style` prop is not as elegant as leveraging Tailwind CSS classes, it guarantees the desired functionality. Further investigation may be required to determine why the dynamic class approach fails and whether a more integrated solution within the Tailwind CSS framework exists.

---

Please feel free to reach out if any questions or further clarifications are needed.

**Comment by wongjn** [comment]
As per [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as _complete unbroken strings_ in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:
>
> **Don’t construct class names dynamically**
>
> ```vue
> <div class="text-{{ error ? 'red' : 'green' }}-600"></div>
> ```
> In the example above, the strings `text-red-600` and `text-green-600` do not exist, so Tailwind will not generate those classes.
> Instead, make sure any class names you’re using exist in full:
>
> **Always use complete class names**
>
> ```vue
> <div class="{{ error ? 'text-red-600' : 'text-green-600' }}"></div>
> ```
